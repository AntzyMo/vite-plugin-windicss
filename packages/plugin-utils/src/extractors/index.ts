export * from './default'
export * from './helper'
export * from './pug'
export * from './svelte'
