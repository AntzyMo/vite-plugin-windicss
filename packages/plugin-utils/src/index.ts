export * from './constants'
export * from './defineConfig'
export * from './createUtils'
export * from './options'
export * from './transforms'
export * from './extractors'
export * from './utils'
export * from './resolveOptions'
export * from './merge'
