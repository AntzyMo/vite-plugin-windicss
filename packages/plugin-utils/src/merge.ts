import type { FullConfig } from 'windicss/types/interfaces'

const isObject = (val: any) => toString.call(val) === '[object Object]'

function deepMerge(a: any, b: any, rootPath: string) {
  a = { ...a }
  Object.keys(b).forEach((key) => {
    if (isObject(a[key]))
      a[key] = deepMerge(a[key], b[key], rootPath ? `${rootPath}.${key}` : key)
    else if (Array.isArray(a[key]))
      a[key] = [...a[key], ...b[key]]
    else
      a[key] = b[key]
  })
  return a
}

export function mergeWindicssConfig(a: FullConfig, b: FullConfig) {
  // TODO: handle more special props
  return deepMerge(a, b, '')
}
