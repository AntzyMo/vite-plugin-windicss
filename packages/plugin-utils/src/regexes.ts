export const regexQuotedString = /(["'`])((?:\\\1|(?:(?!\1)|\n|\r).)*?)\1/mg
export const regexClassCheck = /^[!a-z\d.+-]+[\w:/\\.$()-]*[\w)]$/
export const regexHtmlTag = /<(\w[\w-]*)[\S\s]*?\/?>/mg
export const regexClassSplitter = /[\s'"`{}]/g
export const regexClassGroup = /([!\w+-][\w+:_/-]*?\w):\(([\w\s/-]*?)\)/gm
