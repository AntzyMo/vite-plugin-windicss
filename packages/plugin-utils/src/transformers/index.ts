export * from './pug'
export * from './types'
