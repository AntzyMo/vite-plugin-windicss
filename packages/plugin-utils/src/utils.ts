import { toArray } from '@antfu/utils'

export type { Arrayable } from '@antfu/utils'
export { toArray, partition, slash } from '@antfu/utils'

export type NestedArrayable<T> = T | (T | T[])[]

export function flattenArray<T>(v: NestedArrayable<T>): T[] {
  return toArray(v).flat() as T[]
}

export function mergeArrays<T>(...args: (NestedArrayable<T> | undefined)[]): T[] {
  return args.flatMap(i => flattenArray(i || []))
}

export function kebabCase(str: string) {
  return str.replace(/([a-z])([A-Z])/g, '$1-$2').toLowerCase()
}

export function include<T>(set: Set<T>, v: T[] | Set<T>) {
  for (const i of v)
    set.add(i)
}

export function exclude<T>(set: Set<T>, v: T[] | Set<T>) {
  for (const i of v)
    set.delete(i)
}

export function escapeRegExp(str: string) {
  return str.replace(/[.*+?^${}()|[\]\\]/g, '\\$&') // $& means the whole matched string
}
