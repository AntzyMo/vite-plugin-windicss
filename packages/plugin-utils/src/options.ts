/* eslint-disable no-use-before-define */

import type { Config as WindiCssOptions } from 'windicss/types/interfaces'
import { defaultAlias, TagNames } from './constants'
import { kebabCase, toArray } from './utils'

export { WindiCssOptions }

export interface WindiPluginUtilsOptions {
  /**
   * Options for windicss/tailwindcss.
   * Also accepts string as config file path.
   *
   * @default 'tailwind.config.js'
   */
  config?: WindiCssOptions | string

  /**
   * Safe class names to be always included.
   */
  safelist?: string | string[]

  /**
   * Enabled windicss preflight (a.k.a TailwindCSS style reset)
   *
   * @default true
   */
  preflight?: boolean | {
    /**
     * Enable all the preflight regardless the template
     */
    enableAll?: boolean

    /**
     * Safelist to always included
     */
    safelist?: string | string[]

    /**
      * Alias for resolving preflight
      */
    alias?: Record<string, TagNames>

    /**
     * @default true
     */
    includeBase?: boolean

    /**
     * @default true
     */
    includeGlobal?: boolean

    /**
     * @default true
     */
    includePlugin?: boolean
  }

  /**
   * Scan the files and extract the usage
   *
   * @default true
   */
  scan?: boolean | {
    /**
     * Auto scan on startup
     *
     * @default true
     */
    runOnStartup?: boolean

    /**
     * Directories to search for classnames
     *
     * @default 'src'
     */
    dirs?: string[]

    /**
     * File extension to search for classnames
     *
     * @default 'html', 'vue'
     */
    fileExtensions?: string[]

    /**
     * Exclude globs
     *
     * @default []
     */
    exclude?: string[]

    /**
     * Include globs
     *
     * @default []
     */
    include?: string[]
  }

  /**
   * Transform CSS for `@apply` directive
   *
   * @default true
   */
  transformCSS?: boolean

  /**
   * Transform groups like `hover:(bg-gray-100 font-medium)`
   *
   * @default true
   */
  transformGroups?: boolean

  /**
   * Sort the genrate utilities
   *
   * @default true
   */
  sortUtilities?: boolean

  /**
   * Name for debug
   *
   * @default 'windi-plugin-utils'
   * @internal
   */
  _pluginName?: string

  /**
   * CWD
   *
   * @default process.cwd
   * @internal
   */
  _projectRoot?: string
}

export type UserOptions = Omit<WindiPluginUtilsOptions, '_pluginName' | '_projectRoot'>

export function resolveOptions(options: WindiPluginUtilsOptions) {
  const {
    config = 'tailwind.config.js',
    scan = true,
    preflight = true,
    transformCSS = true,
    sortUtilities = true,
    _projectRoot = process.cwd(),
    _pluginName = 'windi-plugin-utils',
  } = options

  const preflightOptions = Object.assign(
    {
      includeBase: true,
      includeGlobal: true,
      includePlugin: true,
      enableAll: false,
      alias: {},
    },
    typeof preflight === 'boolean' ? {} : preflight,
  )

  const scanOptions = Object.assign(
    {
      fileExtensions: ['html', 'vue', 'md', 'pug', 'jsx', 'tsx', 'svelte'],
      dirs: ['src'],
      exclude: [] as string[],
      include: [] as string[],
      runOnStartup: true,
    },
    typeof scan === 'boolean' ? {} : scan,
  )

  const safelist = new Set(toArray(options.safelist || []).flatMap(i => i.split(' ')))

  preflightOptions.alias = Object.fromEntries(
    Object.entries({
      ...defaultAlias,
      ...preflightOptions.alias,
    }).filter(([k, v]) => [kebabCase(k), v]),
  )

  return {
    ...options,
    _projectRoot,
    _pluginName,
    config,
    scan: Boolean(scan),
    scanOptions,
    preflight: Boolean(preflight),
    preflightOptions,
    transformCSS,
    sortUtilities,
    safelist,
  }
}

export type ResolvedOptions = ReturnType<typeof resolveOptions>
