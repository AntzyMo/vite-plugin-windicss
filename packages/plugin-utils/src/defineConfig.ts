import { WindiCssOptions } from './options'

export function defineConfig(config: WindiCssOptions) {
  return config
}
