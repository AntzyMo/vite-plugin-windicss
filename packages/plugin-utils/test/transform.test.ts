import { createUtils } from '../src'

describe('transfrom', () => {
  it('groups', async() => {
    const cases = [
      'bg-white font-light sm:hover:(bg-gray-100 font-medium)',
      '-sm:hover:(p-1 p-2)',
      '+sm:(p-1 p-2)',
      'dark:+lg:(p-1 p-2)',
    ]
    const utils = createUtils()

    for (const c of cases) {
      const transformed = utils.transformGroups(c)
      expect(transformed).toMatchSnapshot(`"${c}"`)
      const transformedSourceMap = utils.transformGroupsWithSourcemap(c)?.code
      expect(transformedSourceMap).toBe(transformed)
    }
  })
})
