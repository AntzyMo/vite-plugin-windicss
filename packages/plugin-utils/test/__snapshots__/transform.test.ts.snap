// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`transfrom groups: "+sm:(p-1 p-2)" 1`] = `"+sm:p-1 +sm:p-2"`;

exports[`transfrom groups: "-sm:hover:(p-1 p-2)" 1`] = `"-sm:hover:p-1 -sm:hover:p-2"`;

exports[`transfrom groups: "bg-white font-light sm:hover:(bg-gray-100 font-medium)" 1`] = `"bg-white font-light sm:hover:bg-gray-100 sm:hover:font-medium"`;

exports[`transfrom groups: "dark:+lg:(p-1 p-2)" 1`] = `"dark:+lg:p-1 dark:+lg:p-2"`;
